applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_local_stress_response_function.h

```diff
@@ -0,0 +1,599 @@
+
+			FiniteDifferencesUtilities::Pointer FD_calculate_gradient_2 = m_traced_pElement->GetValue(FINITE_DIFFERENCE_INFORMATION);
+
+			std::cout << "Test finite difference stuff = " << FD_calculate_gradient_2->GetDesignVariable() << std::endl;*/
+			//------------------------------
+
+			Matrix stress_displ_deriv;
+			if(m_stress_treatment == "mean" || m_stress_treatment == "GP")
+				m_traced_pElement->Calculate(STRESS_DISP_DERIV_ON_GP, stress_displ_deriv, rProcessInfo);
+			else
+				m_traced_pElement->Calculate(STRESS_DISP_DERIV_ON_NODE, stress_displ_deriv, rProcessInfo);
+
+			int num_of_dofs = stress_displ_deriv.size1();
+			int num_of_deriv = stress_displ_deriv.size2();
+			double stress_displ_deriv_value = 0.0;
+
+			if(rResponseGradient.size() != stress_displ_deriv.size1())
+				KRATOS_ERROR << "Size of stress displacement derivative does not fit!" << std::endl;
+
+			for (int dof_it = 0 ; dof_it < num_of_dofs; dof_it++)
+			{
+				if(m_stress_treatment == "mean")
+				{
+					for(int GP_it = 0; GP_it < num_of_deriv; GP_it++)
+						stress_displ_deriv_value += stress_displ_deriv(dof_it, GP_it);
+
+					stress_displ_deriv_value /= num_of_deriv;
+				}
+				else if(m_stress_treatment == "GP")
+				{
+					if(num_of_deriv >= m_id_of_location)
+						stress_displ_deriv_value = stress_displ_deriv(dof_it, (m_id_of_location-1));
+					else
+						KRATOS_ERROR << "Chosen Gauss-Point is not availible. Chose 'stress_location' between 1 and " <<
+									num_of_deriv  << "!"<< std::endl;
+				}
+				else if(m_stress_treatment == "node")
+				{
+					if(num_of_deriv >= m_id_of_location)
+						stress_displ_deriv_value = stress_displ_deriv(dof_it, (m_id_of_location-1));
+					else
+						KRATOS_ERROR << "Chosen node is not availible. The element has only " <<
+									num_of_deriv  << " nodes."<< std::endl;
+
+				}
+				rResponseGradient[dof_it] = (-1) * stress_displ_deriv_value;
+				//std::cout << ("stress_disp_deriv_value = ") << -1*stress_displ_deriv_value << std::endl;
+				stress_displ_deriv_value = 0.0;
+			}
+		}
+	}
+
+
+protected:
+	///@name Protected static Member Variables
+	///@{
+
+	///@}
+	///@name Protected member Variables
+	///@{
+
+	///@}
+	///@name Protected Operators
+	///@{
+
+	///@}
+	///@name Protected Operations
+	///@{
+
+	// ==============================================================================
+	void CalculateSensitivityGradient(Element& rAdjointElem,
+                                      const Variable<double>& rVariable,
+                                      const Matrix& rDerivativesMatrix,
+                                      Vector& rResponseGradient,
+                                      ProcessInfo& rProcessInfo) override
+    {
+      	KRATOS_TRY
+
+
+		if(rAdjointElem.Id() == m_id_of_traced_element)
+		{
+			rAdjointElem.SetValue(DESIGN_VARIABLE_NAME, rVariable.Name());
+
+			Matrix stress_DV_deriv;
+			if(m_stress_treatment == "mean" || m_stress_treatment == "GP")
```

Comment:
i really would not do string checking on a lement by element level.

can't enums be definefor stress_treatment?